tikv/mvcc.go

```diff
@@ -119,22 +129,108 @@ func (store *MVCCStore) updateLatestTS(ts uint64) {
+	regCtx := reqCtx.regCtx
+	hashVals := mutationsToHashVals(mutations)
+	regCtx.acquireLatches(hashVals)
+	defer regCtx.releaseLatches(hashVals)
+
+	// check again after acquired latch.
+	dup, err := store.checkPessimisticLock(reqCtx, mutations, startTS)
+	if dup || err != nil {
+		return store.handleCheckPessimisticErr(startTS, err)
+	}
+
+	lockBatch := newWriteLockBatch(reqCtx)
+	// Check the DB.
+	txn := reqCtx.getDBReader().GetTxn()
+	keys := make([][]byte, len(mutations))
+	for i, m := range mutations {
+		keys[i] = m.Key
+	}
+	items, err := txn.MultiGet(keys)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, m := range mutations {
+		// Note the use of `forUpdateTS` here!
+		oldMeta, oldVal, err := store.checkConflictInDB(reqCtx, txn, items[i], forUpdateTS)
+		if err != nil {
+			return nil, err
+		}
+		lock := mvcc.MvccLock{
+			MvccLockHdr: mvcc.MvccLockHdr{
+				StartTS:    startTS,
+				Op:         uint8(kvrpcpb.Op_PessimisticLock),
+				HasOldVer:  oldMeta != nil,
+				TTL:        uint32(ttl),
+				PrimaryLen: uint16(len(primary)),
+			},
+			Primary: primary,
+			Value:   m.Value,
+			OldMeta: oldMeta,
+			OldVal:  oldVal,
+		}
+		lockBatch.set(m.Key, lock.MarshalBinary())
+	}
+	err = store.writeLocks(lockBatch)
+	log.Info(startTS, "locked", hashVals)
+	return nil, err
+}
+
+func (store *MVCCStore) handleCheckPessimisticErr(startTS uint64, err error) (*lockwaiter.Waiter, error) {
+	if errLock, ok := err.(*ErrLocked); ok {
+		waiter := store.lockWaiterManager.NewWaiter(startTS, errLock.StartTS, farm.Fingerprint64(errLock.Key), time.Second)
+		return waiter, err
+	}
+	return nil, err
+}
+
+func (store *MVCCStore) checkPessimisticLock(reqCtx *requestCtx, mutations []*kvrpcpb.Mutation, startTS uint64) (dup bool, err error) {
+	for _, m := range mutations {
+		lock, err := store.checkConflictInLockStore(reqCtx, m, startTS)
+		if err != nil {
+			return false, err
+		}
+		if lock != nil {
+			// The lock already exists, it's a duplicate command, we can return directly.
+			y.Assert(lock.Op == uint8(kvrpcpb.Op_PessimisticLock))
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (store *MVCCStore) Prewrite(reqCtx *requestCtx, req *kvrpcpb.PrewriteRequest) []error {
+	mutations := req.Mutations
+	startTS := req.StartVersion
+	primary := req.PrimaryLock
+	ttl := req.LockTtl
 	regCtx := reqCtx.regCtx
 	hashVals := mutationsToHashVals(mutations)
 	errs := make([]error, 0, len(mutations))
 	anyError := false
 
 	regCtx.acquireLatches(hashVals)
 	defer regCtx.releaseLatches(hashVals)
-
+	lockBatch := newWriteLockBatch(reqCtx)
+	hasPessimistic := len(req.PessimisticLockExist) > 0
 	// Must check the LockStore first.
-	for _, m := range mutations {
-		duplicate, err := store.checkPrewriteInLockStore(reqCtx, m, startTS)
+	for i, m := range mutations {
+		lock, err := store.checkConflictInLockStore(reqCtx, m, req.StartVersion)
 		if err != nil {
 			anyError = true
 		}
-		if duplicate {
+		if hasPessimistic && req.PessimisticLockExist[i] {
```

Comment:
What's the use of req.PessimisticLockExist[i]? Only all pessimistic locks exist, prewrite can succeed.

Reply:
non-unique index and delete index don't need to be locked.